### PR TITLE
added filtersForm

### PR DESCRIPTION
**Pull request by YehudaGold**
This PR add the ability to filter on branches proprieties in the pages.

- Added ` FiltersForm` slot component to the `topbar`.
- Added blue and yellow `useFilters` hooks that return the filtered branches and the  ` FiltersForm` component.
- The filter and search logic moved to the main page component to join the logic .
- Fixed `useBranchesFilter`` output not re-calculated on input branches change.

### CODE REVIEW

File: hooks/useBooleanInput.tsx [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { FormControlLabel, Switch } from "@mui/material";
```
Comment (nivzelber) [inline hooks/useBooleanInput.tsx]: put all form inputs under `hooks/form`

File: hooks/useListInput.tsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import { Autocomplete, TextField } from "@mui/material";
+import { uniq } from "lodash";
```
Comment (nivzelber) [inline hooks/useListInput.tsx]: add `lodash` as a dependency

File: hooks/useListInput.tsx [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import { useState, useMemo } from "react";
+
+export const useListInput = <TBranch extends Record<string, string>>(branches: TBranch[], key: keyof TBranch, label: string ) => {
```
Comment (nivzelber) [inline hooks/useListInput.tsx]: This component can be more general.
It only works with branches right now.

At the bare minimum rename it to `useBranchPropertyListInput`

File: hooks/useListInput.tsx [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  const options = useMemo(
+        () => uniq(branches.map(branch => branch[key])).sort(),
+        [branches, key]
+    );
```
Comment (nivzelber) [inline hooks/useListInput.tsx]: add an empty line here

File: hooks/useListInput.tsx [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    const handleCategoryChange = (_: unknown, newValue: string | null) => {
+      setValue(newValue);
+    };
+
+    const Input = (
+         <Autocomplete
+            onChange={handleCategoryChange}
+            options={options}
+            value={value}
+            renderInput={(params) => <TextField {...params} label={label} sx={{padding:"0.5em"}}/>}
```
Comment (nivzelber) [inline hooks/useListInput.tsx]:
```suggestion
            renderInput={(props) => <TextField {...props} label={label} sx={{padding:"0.5em"}}/>}
```

File: components/pages/blue/useFilters.tsx [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            if (area) filteredBranches = filteredBranches.filter(branch => branch.area === area);
+            if (city) filteredBranches = filteredBranches.filter(branch => branch.city === city);
+            
+            return filteredBranches;
+        },
+        [area, branches, category, city, delivery, kosher]
+    );
+
+    const FiltersForm = (
+        <FormGroup sx={{width:"18em"}}>
```
Comment (nivzelber) [inline components/pages/blue/useFilters.tsx]: 18 is no magic number?
This has to be mobile and desktop compatiable.

File: components/pages/yellow/useFilters.tsx [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+            if (isOnline) filteredBranches = filteredBranches.filter(branch => branch.is_online === "Y");
+            if (category) filteredBranches = filteredBranches.filter(branch => branch.company_category === category);
+                
+            return filteredBranches;
+        },
+        [branches, category, isOnline]
+    );
+
+    const FiltersForm = (
+        <FormGroup sx={{width:"13em"}}>
```
Comment (nivzelber) [inline components/pages/yellow/useFilters.tsx]: same with magic number

File: components/layout/topbar/topbar.styled.tsx [hunk 8]
```diff
@@ -40,3 +40,7 @@ export const StyledInputBase = styled(InputBase)`
     padding-right: calc(1em + 10px);
   }
 `;
+
+export const Space = styled.div`
```
Comment (nivzelber) [inline components/layout/topbar/topbar.styled.tsx:8]: What does this does? I've seen the usage but cannot figure it out...

**Review by nivzelber — COMMENTED** [review]
Great work! Awesome changes!
Please do the following:
1. Pull `master` again (further changes were introduced)
2. Fix the PR
3. Submit again and I'll approve

File: components/layout/topbar/topbar.styled.tsx (hunk 8 above)
Comment (YehudaGold) [reply components/layout/topbar/topbar.styled.tsx:8]: This `div` added space in the nav-bar by growing in the flex container, for moving all elements after it to the left corner.

In the documentation they put it on a `Typography` field to push everything. 
[ButtonAppBar](https://mui.com/material-ui/react-app-bar/#ButtonAppBar.tsx)

File: components/pages/blue/useFilters.tsx (hunk 6 above)
Comment (YehudaGold) [reply components/pages/blue/useFilters.tsx]: I can transfer it to a new style-component, maybe round it, but in the end yes - this is the size I guess the form of filters needs, I assume it by the input of categories.

Like `200px` in the search bar.

File: components/pages/yellow/useFilters.tsx (hunk 7 above)
Comment (YehudaGold) [reply components/pages/yellow/useFilters.tsx]: [the same comment](https://github.com/NIV54/hvr-search/pull/1/files#r1161153901)

File: hooks/useListInput.tsx (hunk 3 above)
Comment (YehudaGold) [reply hooks/useListInput.tsx]: I tried to make the name more generic, it can run on any list of objects.
I updated the parameter names.

**Comment by YehudaGold** [comment]
I updated my stuff, and answered what I could, I'd love to know what else I can improve.

**Review by nivzelber — APPROVED** [review]

**Pull request merged**